File: sic-app/src/app/feature/pm/dt/pmdt07/change-request.service.ts
// src/app/core/services/change-request.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';
import { environment } from '../../../../../environments/environment';
import type { ChangeRequest, ImpactAnalysis } from './change-request.model';


import { PaginationResponse } from '../../../../core/model/sic-base-model';

@Injectable({ providedIn: 'root' })
export class ChangeRequestService {
  private http = inject(HttpClient);
  private baseUrl = environment.apiBaseUrl + '/api/pm/change-requests'; // สมมติ endpoint
  private impactUrl = environment.apiBaseUrl + '/api/pm/impact-analysis';

  // ===== Change Request CRUD =====
  getChangeRequests(params?: {
    projectId?: string;
    requirementId?: string;
    status?: string;
    keyword?: string;
    page?: number;
    size?: number;
  }): Observable<PaginationResponse<ChangeRequest>> {
    let httpParams = new HttpParams();
    if (params) {
      Object.entries(params).forEach(([key, value]) => {
        if (value !== undefined && value !== null) {
          httpParams = httpParams.set(key, String(value));
        }
      });
    }
    return this.http.get<PaginationResponse<ChangeRequest>>(this.baseUrl, { params: httpParams });
  }

  getChangeRequest(id: string): Observable<ChangeRequest> {
    return this.http.get<ChangeRequest>(`${this.baseUrl}/${id}`);
  }

  saveChangeRequest(data: ChangeRequest): Observable<string> {
    return this.http.post<string>(`${this.baseUrl}/save`, data);
  }

  deleteChangeRequest(id: string): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${id}`);
  }

  // ===== Impact Analysis =====
  getImpactByChangeRequest(changeRequestId: string): Observable<ImpactAnalysis> {
    return this.http.get<ImpactAnalysis>(`${this.impactUrl}/change-request/${changeRequestId}`);
  }

  autoDetectImpact(changeRequestId: string): Observable<ImpactAnalysis> {
    return this.http.post<ImpactAnalysis>(`${this.impactUrl}/auto-detect/${changeRequestId}`, {});
  }

  saveImpactAnalysis(data: ImpactAnalysis): Observable<string> {
    return this.http.post<string>(`${this.impactUrl}/save`, data);
  }

  deleteImpactAnalysis(id: string): Observable<void> {
    return this.http.delete<void>(`${this.impactUrl}/${id}`);
  }
  // ใน change-request.service.ts เพิ่ม
  createChangeRequest(data: any): Observable<any> {
    return this.http.post(`${this.baseUrl}`, data);
  }
  submitForApproval(id: string): Observable<any> {
    return this.http.post(`${this.baseUrl}/${id}/submit`, {});
  }
  approve(id: string): Observable<any> {
    return this.http.post(`${this.baseUrl}/${id}/approve`, {});
  }
  reject(id: string, reason?: string): Observable<any> {
    return this.http.post(`${this.baseUrl}/${id}/reject`, { reason });
  }
  implement(id: string): Observable<any> {
    return this.http.post(`${this.baseUrl}/${id}/implement`, {});
  }
  markAssigneeComplete(id: string, userId: string, targetId: string): Observable<any> {
    return this.http.post(`${this.baseUrl}/${id}/assignees/complete`, {}, { params: { userId, targetId } });
  }
  checkEditLock(targetType: string, targetId: string): Observable<any> {
    return this.http.get(`${environment.apiBaseUrl}/api/pm/edit-sessions/check`, { params: { targetType, targetId } });
  }
}

File: sic-spring/sic/src/main/java/com/softinter/sicapi/service/ChangeRequestService.java
package com.softinter.sicapi.service;

import com.softinter.sicapi.dto.request.ChangeRequestRequest;
import com.softinter.sicapi.dto.response.ChangeRequestResponse;
import com.softinter.sicapi.dto.response.PaginationResponse;
import org.springframework.data.domain.Pageable;

import java.util.UUID;

public interface ChangeRequestService {
    ChangeRequestResponse createChangeRequest(ChangeRequestRequest request);
    ChangeRequestResponse updateChangeRequest(UUID id, ChangeRequestRequest request);
    ChangeRequestResponse getChangeRequest(UUID id);
    PaginationResponse<ChangeRequestResponse> listChangeRequests(UUID projectId, String targetType, UUID targetId, String status, String keyword, Pageable pageable);
    void deleteChangeRequest(UUID id);

    /** ตัวเลือกเอกสารเป้าหมายของ CR: เฉพาะเอกสารที่อนุมัติแล้ว (ถ้าระบุ value คืนรายการเดียวเพื่อแสดงค่าที่เลือกไว้เดิม) */
    java.util.List<com.softinter.sicapi.dto.response.ComboboxResponse> getApprovedTargetCombobox(
            String targetType, UUID projectId, String keyword, UUID value);
    ChangeRequestResponse submitForApproval(UUID id);
    ChangeRequestResponse approve(UUID id, String approvedBy);
    ChangeRequestResponse reject(UUID id, String reason);
    ChangeRequestResponse implement(UUID id);
    ChangeRequestResponse markAssigneeComplete(UUID changeRequestId, String userId, UUID targetId);
}